src/varnam_state.cpp

```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
 }
 
 void VarnamState::commitText(const FcitxKeySym &key) {
-  auto candidates = ic_->inputPanel().candidateList();
-  std::string stringToCommit;
-
-  if (key == FcitxKey_Escape || key == FcitxKey_0 || candidates == nullptr ||
-      candidates->size() <= 1 || result_ == nullptr ||
-      varray_is_empty(result_) || lastTypedCharIsDigit) {
-
-    stringToCommit.assign(preedit_.toStringForCommit());
-    candidateSelected = 0;
-  } else if ((candidates->cursorIndex() <= 0) && !candidateSelected) {
-    vword *first_result = static_cast<vword *>(varray_get(result_, 0));
-    if (first_result != nullptr) {
-      stringToCommit.assign(first_result->text);
-      candidateSelected = 1;
+    auto candidates = ic_->inputPanel().candidateList();
+    std::string stringToCommit;
+    std::string wordToLearn;
+    bool isWordBreakKey = isWordBreak(key);
+    bool enablePunctuation = engine_->getConfig()->enablePunctuation.value();
+
+    if (isWordBreakKey) {
+        if (candidates && candidates->cursorIndex() >= 0) {
+            stringToCommit = candidates->candidate(candidates->cursorIndex()).text().toStringForCommit();
+        } else {
+            stringToCommit = preedit_.toStringForCommit();
+        }
+
+        if (enablePunctuation) { // If punctuation processing is enabled
+            buffer_.clear(); // Clear buffer
+            buffer_.push_back(getWordBreakChar(key)[0]); // Add word break char to buffer
+            
+            // Concatenate result word if available
+            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
+                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
+                if (result_word) {
+                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
+                }
+            }
+        } else {
+            // If punctuation processing is not enabled
+            // Append the punctuation character to stringToCommit
+            stringToCommit += getWordBreakChar(key);
```

Comment:
always use stringstream or similar API to concatenate std::string, otherwise it would cause unnecessary memory allocations.

Reply:
On a second thought, don't worry about it. It wouldn't make much difference, its just two strings.

Reply:
I did use sstream on another branch here  https://github.com/khumnath/varnam-fcitx5/blob/2cf6dc506d39a235c472b3e416bebd48d029160c/src/varnam_state.cpp#L370-L444 .
 can i pull it on dev branch?

Reply:
We can go with std::string, no need to use std::stringstream here

Reply:
if we can use std::string then this is solved i think.